### PR TITLE
Test rework, part 4

### PR DESCRIPTION
**Pull request by apostasie**
This is the next part of the test rewriting project.

### Background and motivation:

See:
- #3418
- #3455
- #3464 

and the initial commit message from https://github.com/containerd/nerdctl/pull/3418/commits/2c95d090237c534b9ddec59b6488386445c41da6
along with the linked issues from above PRs.

The gist of it is really simple: our testing situation is not good (see prior discussions for full list of issues - but then, the sheer fact we have to retry our CI multiple times to get a green tree should say enough).

This here is the part 4 of the effort to rewrite our tests using a new, custom developed testing toolkit that emphasizes test expressiveness, debugability and better isolation.

### Changes

Individual commits messages have extensive details about the changes where appropriate.

### Status

_a. the following has been ran locally 100x+ with no failure:_
  - . (< 1 sec)
  - completion (< 1 sec)
  - issues (< 10 secs)
  - system (~ 5 secs)
  - volume (~ 5 secs)
  - network  (~ 5 secs)
    - although, #3500 did show up and might still be lurking (~~and #3501~~, albeit this one probably got fixed by #3508)

_b. the following have ran locally 20x+ with no failure:_
  - ipfs (~ 100 secs)

_c. the following are failing after about 20 iterations:_
  - image (~40 secs)
    - likely due to one of the variants in #3513 (the one with `images --filter` or the one using a `platform reduced` push)

_d. the following are NOT part of this PR:_
  - builder
  - container
  - compose
  - login

Also left out of this PR: disabling retries for non flaky tests.
We do now have the infrastructure for that (eg: `-test.only-flaky`) but this will come in another PR after this.

Note that while no-failure locally with 100+ runs is no guarantee that things are fine: I do run local tests directly on the host, not inside the dockerfile - running inside the dockerfile on the CI is likely to bring up a lot more failures, either because things are inherently slower, or because the extra docker layer is adding specific conditions that will make us fail.

---------------------------------------------

### Review

Close to 10k lines is indeed a chunky PR, but then, as with the previous parts, there is absolutely no "code" or "logic" changes in nerdctl itself - this is "just" tests: if they run, and if the CI is green...

PR has been split in as many commits as feasible to facilitate review, along with a number of inline comments in the github UI where appropriate.

Let me know if I can make this easier in any way.

The north star for this whole effort remains:
- we must get to a **_consistently green CI without retries - with no or very few flaky test_**.
- `go test -p 1 ./cmd/nerdctl/...` must work locally out of the box on developers laptop without randomly breaking for missing requirements

### CODE REVIEW

File: cmd/nerdctl/volume/volume_create_test.go [hunk 1]
```diff
@@ -27,27 +28,27 @@ import (
...
+	testCase := nerdtest.Setup()
 
-	tg := &test.Group{
+	testCase.SubTests = []*test.Case{
 		{
 			Description: "arg missing should create anonymous volume",
-			Command:     test.RunCommand("volume", "create"),
-			Expected:    test.Expects(0, nil, nil),
+			Command:     test.Command("volume", "create"),
+			Expected:    test.Expects(0, nil, test.Match(regexp.MustCompile("^[a-f0-9]{64}\n$"))),
```
Comment (apostasie) [inline cmd/nerdctl/volume/volume_create_test.go:37]: Enhance the test a little bit.

File: cmd/nerdctl/volume/volume_create_test.go [hunk 2]
```diff
@@ -74,28 +75,23 @@ func TestVolumeCreate(t *testing.T) {
...
 				helpers.Anyhow("volume", "rm", "-f", data.Identifier())
 			},
-			Expected: func(data test.Data, helpers test.Helpers) *test.Expected {
-				return &test.Expected{
-					// NOTE: docker returns 125 on this
-					ExitCode: -1,
-					Errors:   []error{errdefs.ErrInvalidArgument},
-				}
-			},
+			// NOTE: docker returns 125 on this
```
Comment (apostasie) [inline cmd/nerdctl/volume/volume_create_test.go:86]: Use short syntax.

File: cmd/nerdctl/volume/volume_inspect_test.go [hunk 3]
```diff
@@ -42,162 +42,170 @@ func createFileWithSize(mountPoint string, size int64) error {
...
-			helpers.Ensure("volume", "create", data.Identifier())
-			helpers.Ensure("volume", "create", "--label", "foo=fooval", "--label", "bar=barval", data.Identifier()+"-second")
-			// Obviously note here that if inspect code gets totally hosed, this entire suite will
-			// probably fail right here on the Setup instead of actually testing something
-			vol := nerdtest.InspectVolume(helpers, data.Identifier())
-			err := createFileWithSize(vol.Mountpoint, size)
-			assert.NilError(t, err, "File creation failed")
+	testCase := nerdtest.Setup()
+
+	testCase.Require = nerdtest.BrokenTest("This test assumes that the host-side of a volume can be written into, "+
```
Comment (apostasie) [inline cmd/nerdctl/volume/volume_inspect_test.go:49]: Marking this test as broken (for Docker without root)

File: cmd/nerdctl/volume/volume_list_test.go [hunk 4]
```diff
@@ -77,313 +81,320 @@ func TestVolumeLsSize(t *testing.T) {
...
 func TestVolumeLsFilter(t *testing.T) {
-	nerdtest.Setup()
+	testCase := nerdtest.Setup()
 
-	tc := &test.Case{
-		Description: "Volume ls",
-		Setup: func(data test.Data, helpers test.Helpers) {
-			var vol1, vol2, vol3, vol4 = data.Identifier() + "-1", data.Identifier() + "-2", data.Identifier() + "-3", data.Identifier() + "-4"
-			var label1, label2, label3, label4 = data.Identifier() + "=label-1", data.Identifier() + "=label-2", data.Identifier() + "=label-3", data.Identifier() + "-group=label-4"
+	testCase.Require = nerdtest.BrokenTest("This test assumes that the host-side of a volume can be written into, "+
```
Comment (apostasie) [inline cmd/nerdctl/volume/volume_list_test.go:92]: Marking test as broken.

File: cmd/nerdctl/main_test_test.go [hunk 5]
```diff
@@ -92,56 +91,15 @@ func TestTest(t *testing.T) {
...
-						return helpers.CustomCommand("printf", data.Get("status"))
+					Command: func(data test.Data, helpers test.Helpers) test.TestableCommand {
+						return helpers.Custom("printf", data.Get("status"))
 					},
 					Expected: test.Expects(0, nil, test.Equals("uninitialized-setup-command")),
 				},
 			},
 			Expected: test.Expects(0, nil, test.Equals("uninitialized-setup")),
 		},
-		{
```
Comment (apostasie) [inline cmd/nerdctl/main_test_test.go:103]: WithEnv is no longer exposed. This tests are now moot.

File: cmd/nerdctl/main_test.go [hunk 6]
```diff
@@ -33,100 +33,98 @@ func TestMain(m *testing.M) {
...
-			{
-				Description: "Default",
-				Command:     test.RunCommand("info", "-f", "{{.Driver}}"),
-				Expected:    test.Expects(0, nil, test.Equals(defaults.DefaultSnapshotter+"\n")),
-			},
-			{
-				Description: "TOML > Default",
-				Command:     test.RunCommand("info", "-f", "{{.Driver}}"),
-				Expected:    test.Expects(0, nil, test.Equals("dummy-snapshotter-via-toml\n")),
-				Data:        test.WithConfig(nerdtest.NerdctlToml, `snapshotter = "dummy-snapshotter-via-toml"`),
```
Comment (apostasie) [inline cmd/nerdctl/main_test.go:99]: `Config` is now separate from `Data`.

File: cmd/nerdctl/issues/main_linux_test.go [hunk 7]
```diff
@@ -24,35 +24,41 @@ import (
...
-	testGroup := &test.Group{
+	testCase.SubTests = []*test.Case{
 		{
 			Description: "-it --net=host",
 			Require:     test.Binary("unbuffer"),
-			Command: func(data test.Data, helpers test.Helpers) test.Command {
-				return helpers.
-					Command("run", "-it", "--rm", "--net=host", testutil.AlpineImage, "echo", "this was always working").
-					WithWrapper("unbuffer")
+			Command: func(data test.Data, helpers test.Helpers) test.TestableCommand {
```
Comment (apostasie) [inline cmd/nerdctl/issues/main_linux_test.go:40]: Chaining was a mistake.

File: cmd/nerdctl/issues/main_linux_test.go [hunk 8]
```diff
@@ -14,7 +14,7 @@
    limitations under the License.
 */
 
-package main
+package issues
```
Comment (apostasie) [inline cmd/nerdctl/issues/main_linux_test.go:17]: Moving this file out of the root, where it does not belong.

File: cmd/nerdctl/issues/issues_linux_test.go [hunk 9]
```diff
@@ -20,40 +20,20 @@ package issues
...
-func TestMain(m *testing.M) {
-	testutil.M(m)
-}
-
 func TestIssue3425(t *testing.T) {
 	nerdtest.Setup()
 
 	var registry *testregistry.RegistryServer
 
-	var ipfsPath string
```
Comment (apostasie) [inline cmd/nerdctl/issues/issues_linux_test.go:45]: Now embedded inside the Requirement IPFS

File: cmd/nerdctl/issues/issues_linux_test.go [hunk 10]
```diff
@@ -145,13 +125,10 @@ func TestIssue3425(t *testing.T) {
 				Description: "with ipfs",
 				Require: test.Require(
 					nerdtest.Private,
+					nerdtest.IPFS,
 					test.Not(test.Windows),
 					test.Not(nerdtest.Docker),
-					test.Binary("ipfs"),
 				),
-				Env: map[string]string{
```
Comment (apostasie) [inline cmd/nerdctl/issues/issues_linux_test.go:152]: nerdtest.IPFS requirement now takes care of this.

File: cmd/nerdctl/volume/volume_list_test.go [hunk 11]
```diff
@@ -32,30 +32,34 @@ func TestVolumeLsSize(t *testing.T) {
 	nerdtest.Setup()
 
 	tc := &test.Case{
-		Description: "Volume ls --size",
-		Require:     test.Not(nerdtest.Docker),
+		Require: test.Not(nerdtest.Docker),
 		Setup: func(data test.Data, helpers test.Helpers) {
-			helpers.Ensure("volume", "create", data.Identifier()+"-1")
```
Comment (apostasie) [inline cmd/nerdctl/volume/volume_list_test.go:38]: Leverage new `Identifier(s ...string)` function

File: cmd/nerdctl/login/login_linux_test.go [hunk 12]
```diff
@@ -242,11 +242,15 @@ func TestLoginAgainstVariants(t *testing.T) {
 			"token",
 		},
 		// Basic auth, with TLS
-		{
-			80,
-			true,
-			"basic",
-		},
+		/*
```
Comment (apostasie) [inline cmd/nerdctl/login/login_linux_test.go:245]: It is not clear how these tests did pass before - very likely, the simplification in `hosts.toml` generation is the culprit. One way or the other, this is just hitting a pre-existing issue that is being addressed as part of the proposal in #3417 
Disabling test for now, at least until we rewrite them and can flag them with `NerdctlNeedsFixing`

File: cmd/nerdctl/login/login_linux_test.go [hunk 13]
```diff
@@ -258,11 +262,15 @@ func TestLoginAgainstVariants(t *testing.T) {
 			"basic",
 		},
 		// Token auth, with TLS
-		{
-			80,
-			true,
```
Comment (apostasie) [inline cmd/nerdctl/login/login_linux_test.go:263]: ibid

**Comment by apostasie** [comment]
@AkihiroSuda this is not completely green yet.

There is likely one more adjustment needed to get it there.

Also, I would prefer to get #3499 in first - and I am likely to land a few small fixes as well for issues spotted while stress-testing (specifically around the `digest not found` cropping back up).

Nevertheless, given this here is a bit chunky, I think it is fair to get a head start to review asap.

At your convenience as usual.

Thanks.

File: cmd/nerdctl/login/login_linux_test.go (hunk 12 above)
Comment (apostasie) [reply cmd/nerdctl/login/login_linux_test.go:245]: Might be a containerd issue actually. Anyhow, will follow-up with this later on in a separate PR/issue.

File: cmd/nerdctl/helpers/testing.go [hunk 14]
```diff
@@ -30,3 +41,97 @@ func CreateBuildContext(t *testing.T, dockerfile string) string {
 	assert.NilError(t, err)
 	return tmpDir
 }
+
```
Comment (apostasie) [inline cmd/nerdctl/helpers/testing.go:44]: As explained in the commit message, we do currently have a mix of platform specific test files (eg: _platform_test - along with dependencies that may be platform limited too) and tests that call `t.Skip` on certain platforms.

These helpers here are moved around into the main helper file so that we can reduce a little bit the platform specific test files.

Either way, all of that has nothing to do here - unless they are used in live code, they should be ultimately moved to `pkg/testutil` (or subsumed / removed). Work for another PR.

**Comment by apostasie** [comment]
@AkihiroSuda green except:
- the windows reg
- arm64 <- can you poke it?

**Comment by AkihiroSuda** [comment]
Needs rebase

**Comment by apostasie** [comment]
> Needs rebase

Rebased.

I am giving up (temporarily) on the missing content bugs - so, no upcoming PR wrt to that problem for now.
Let's just mark `flaky` whatever test is going to show this particular problem for the time being.

File: cmd/nerdctl/helpers/testing.go [hunk 15]
```diff
@@ -30,3 +41,97 @@ func CreateBuildContext(t *testing.T, dockerfile string) string {
...
+		imageIDs = base.Cmd("images", "--no-trunc", "-a", "-q").OutLines()
+		base.T.Logf("pruning following images: %+v", imageIDs)
+		base.Cmd(append([]string{"rmi", "-f"}, imageIDs...)...).Run()
+	}
+}
+
+func ExtractDockerArchive(archiveTarPath, rootfsPath string) error {
+	if err := os.MkdirAll(rootfsPath, 0755); err != nil {
+		return err
+	}
```
Comment (AkihiroSuda) [inline cmd/nerdctl/helpers/testing.go:89]: Can we use t.TempDir

File: cmd/nerdctl/helpers/testing.go [hunk 16]
```diff
@@ -30,3 +41,97 @@ func CreateBuildContext(t *testing.T, dockerfile string) string {
...
+	workDir, err := os.MkdirTemp("", "extract-docker-archive")
```
Comment (AkihiroSuda) [inline cmd/nerdctl/helpers/testing.go:90]: Can we use t.TempDir

File: pkg/testutil/nerdtest/requirements.go [hunk 17]
```diff
@@ -19,125 +19,293 @@ package nerdtest
...
 
 	"github.com/containerd/nerdctl/v2/pkg/buildkitutil"
 	"github.com/containerd/nerdctl/v2/pkg/inspecttypes/dockercompat"
 	"github.com/containerd/nerdctl/v2/pkg/rootlessutil"
-	"github.com/containerd/nerdctl/v2/pkg/testutil"
+	"github.com/containerd/nerdctl/v2/pkg/testutil/nerdtest/platform"
 	"github.com/containerd/nerdctl/v2/pkg/testutil/test"
 )
 
+var BuildkitHost test.ConfigKey = "bkHost"
```
Comment (AkihiroSuda) [inline pkg/testutil/nerdtest/requirements.go:35]: Should consistently begin with an upper char?

File: cmd/nerdctl/helpers/testing.go (hunk 16 above)
Comment (apostasie) [reply cmd/nerdctl/helpers/testing.go:90]: Yes, all of that stuff inside `cmd/helpers` needs to be cleaned-up / rewritten and likely moved to testutil.
Suggesting we do that in another PR once I ll be done migrating all tests.

File: pkg/testutil/test/utilities.go [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"os"
+)
+
+// IsRoot returns true if we are root... simple
+func IsRoot() bool {
```
Comment (AkihiroSuda) [inline pkg/testutil/test/utilities.go:27]: Why can’t we use pkg/rootlessutil.IsRootless

File: pkg/testutil/testutil.go [hunk 19]
```diff
@@ -560,6 +562,10 @@ func M(m *testing.M) {
 	flag.BoolVar(&flagTestKillDaemon, "test.allow-kill-daemon", false, "enable tests that kill the daemon")
 	flag.BoolVar(&flagTestIPv6, "test.only-ipv6", false, "enable tests on IPv6")
 	flag.BoolVar(&flagTestKube, "test.only-kubernetes", false, "enable tests on Kubernetes")
+	flag.BoolVar(&flagTestFlaky, "test.only-flaky", false, "enable testing of flaky tests only")
```
Comment (AkihiroSuda) [inline pkg/testutil/testutil.go:565]: Is there any way to rather skip flaky tests

File: pkg/testutil/nerdtest/requirements.go (hunk 17 above)
Comment (apostasie) [reply pkg/testutil/nerdtest/requirements.go:35]: Some/most of these config keys are internal only and not meant to be used outside of the package (they are set by `Requirements`).

This one is an exception (for now) as I have not touched much the buildkit stuff - and for the time being it needs to be exported outside of the package.

Will definitely revisit in the next PR when I will do the `builder` tests.

File: pkg/testutil/test/utilities.go (hunk 18 above)
Comment (apostasie) [reply pkg/testutil/test/utilities.go:27]: We could.

This is currently used solely to check for a corner case of Docker.

https://github.com/containerd/nerdctl/blob/192d1cb0d5caab1794b3f0608a0660d64e4fc31e/cmd/nerdctl/volume/volume_inspect_test.go#L54

In that context, rootlessutil.IsRootless seems a bit misleading.

Note that this `IsRoot` needs to go eventually - it is not even proper to test the case in question with Docker - what we really want to test is if Docker Daemon is running root, while the cli is not.

File: pkg/testutil/testutil.go (hunk 19 above)
Comment (apostasie) [reply pkg/testutil/testutil.go:565]: That is the default behavior.
If you do not set `test.only-flaky`, only the `non-flaky` tests tests will run.

Comment (apostasie) [reply pkg/testutil/testutil.go:565]: Next PR will cleanup the github actions and separate flaky (with retry) and non-flaky (no retry)

Comment (AkihiroSuda) [reply pkg/testutil/testutil.go:565]: Could you document it in -help

File: pkg/testutil/test/utilities.go (hunk 18 above)
Comment (AkihiroSuda) [reply pkg/testutil/test/utilities.go:27]: “IsRoot” sounds misleading too. I’d rather suggest just calling os.Geteuid directly

File: pkg/testutil/testutil.go (hunk 19 above)
Comment (apostasie) [reply pkg/testutil/testutil.go:565]: Will do.

Comment (apostasie) [reply pkg/testutil/testutil.go]: Done.

File: pkg/testutil/nerdtest/requirements.go (hunk 17 above)
Comment (apostasie) [reply pkg/testutil/nerdtest/requirements.go:35]: Now I am wondering if I understood what you asked.

Were you instead suggesting the value to be "BkHost"?

I will adjust - actually, will adjust to "BuildkitHost".

File: pkg/testutil/test/utilities.go (hunk 18 above)
Comment (apostasie) [reply pkg/testutil/test/utilities.go:27]: Will amend.

Comment (apostasie) [reply pkg/testutil/test/utilities.go]: Done.

**Comment by apostasie** [comment]
Mildly concerned about the repeat failure on arm - annoying part being it is a retry/timeout, hiding the actual issue :s

**Comment by apostasie** [comment]
There is clearly a deadlock happening in `image` tests, but only for `arm64`.

Debugging right now - marking draft until I find the culprit.

**Comment by apostasie** [comment]
Fixed.

File: cmd/nerdctl/volume/volume_inspect_test.go [hunk 20]
```diff
@@ -51,7 +51,7 @@ func TestVolumeInspect(t *testing.T) {
 		&test.Requirement{
 			Check: func(data test.Data, helpers test.Helpers) (bool, string) {
 				isDocker, _ := nerdtest.Docker.Check(data, helpers)
-				return !isDocker || test.IsRoot(), "docker cli needs to be run as root"
+				return !isDocker || os.Geteuid() == 0, "docker cli needs to be run as root"
```
Comment (apostasie) [inline cmd/nerdctl/volume/volume_inspect_test.go:54]: Based on reviewers feedback.

File: cmd/nerdctl/volume/volume_list_test.go [hunk 21]
```diff
@@ -94,7 +95,7 @@ func TestVolumeLsFilter(t *testing.T) {
 		&test.Requirement{
 			Check: func(data test.Data, helpers test.Helpers) (bool, string) {
 				isDocker, _ := nerdtest.Docker.Check(data, helpers)
-				return !isDocker || test.IsRoot(), "docker cli needs to be run as root"
+				return !isDocker || os.Geteuid() == 0, "docker cli needs to be run as root"
```
Comment (apostasie) [inline cmd/nerdctl/volume/volume_list_test.go:98]: Based on reviewers feedback.

File: pkg/testutil/nerdtest/requirements.go [hunk 22]
```diff
@@ -32,7 +32,7 @@ import (
 	"github.com/containerd/nerdctl/v2/pkg/testutil/test"
 )
 
-var BuildkitHost test.ConfigKey = "bkHost"
+var BuildkitHost test.ConfigKey = "BuildkitHost"
```
Comment (apostasie) [inline pkg/testutil/nerdtest/requirements.go:35]: Based on reviewers feedback.

File: pkg/testutil/testutil.go [hunk 23]
```diff
@@ -562,7 +562,7 @@ func M(m *testing.M) {
 	flag.BoolVar(&flagTestKillDaemon, "test.allow-kill-daemon", false, "enable tests that kill the daemon")
 	flag.BoolVar(&flagTestIPv6, "test.only-ipv6", false, "enable tests on IPv6")
 	flag.BoolVar(&flagTestKube, "test.only-kubernetes", false, "enable tests on Kubernetes")
-	flag.BoolVar(&flagTestFlaky, "test.only-flaky", false, "enable testing of flaky tests only")
```
Comment (apostasie) [inline pkg/testutil/testutil.go]: Based on reviewers feedback.

File: pkg/testutil/nerdtest/ambient.go [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+	return testutil.GetEnableIPv6()
+}
+
+func environmentHasKubernetes() bool {
+	return testutil.GetEnableKubernetes()
+}
+
+func environmentIsForFlaky() bool {
+	return testutil.GetFlakyEnvironment()
+}
```
Comment (AkihiroSuda) [inline pkg/testutil/nerdtest/ambient.go:31]: Why not just call testutil functions directly

File: pkg/testutil/nerdtest/requirements.go [hunk 25]
```diff
@@ -19,125 +19,293 @@ package nerdtest
...
-	// against the host install
-	ret = true
-	mess = "buildkitd is enabled"
-	if testutil.GetTarget() == testutil.Nerdctl {
-		_, err := buildkitutil.GetBuildkitHost(testutil.Namespace)
-		if err != nil {
+}
+
+// RootLess marks a test as suitable only for the rootless environment
+var RootLess = &test.Requirement{
```
Comment (AkihiroSuda) [inline pkg/testutil/nerdtest/requirements.go:135]: mayeb s/RootLess/Rootless/

File: pkg/testutil/nerdtest/utilities.go [hunk 26]
```diff
@@ -25,15 +25,19 @@ import (
 
 	"github.com/containerd/nerdctl/v2/pkg/inspecttypes/dockercompat"
 	"github.com/containerd/nerdctl/v2/pkg/inspecttypes/native"
+	"github.com/containerd/nerdctl/v2/pkg/testutil"
 	"github.com/containerd/nerdctl/v2/pkg/testutil/test"
 )
 
+func IsDocker() bool {
+	return testutil.GetTarget() == "docker"
+}
```
Comment (AkihiroSuda) [inline pkg/testutil/nerdtest/utilities.go:34]: Why not just call testutil.GetTarget directly

**Review by AkihiroSuda — APPROVED** [review]
Thanks, I'm merging this, but this amount of PR was almost unreviewable and even causing Safari to almost hang up 😓 

In the next series of the PR please consider splitting the PR to smaller parts.


Also consider reducing the amount of the helper functions and constants (see the comments)

File: pkg/testutil/nerdtest/ambient.go (hunk 24 above)
Comment (apostasie) [reply pkg/testutil/nerdtest/ambient.go:31]: Currently we have to bridge between the legacy tests, and the code handling flag parsing, and the proposed new framework, which unfortunately implies a certain level of duplication.
I will clean that up once all tests are migrated.

File: pkg/testutil/nerdtest/platform/platform_linux.go [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+func DataHome() (string, error) {
+	return rootlessutil.XDGDataHome()
+}
+
+var (
+	RegistryImageStable = testutil.RegistryImageStable
+	RegistryImageNext   = testutil.RegistryImageNext
+	KuboImage           = testutil.KuboImage
+	DockerAuthImage     = testutil.DockerAuthImage
+)
```
[the remark this answers is not in the document]
Comment (apostasie) [reply pkg/testutil/nerdtest/platform/platform_linux.go:33]: Because we are stuck in a bad situation right now with a mix of `_platform_test.go` files (which should disappear) and `Skip(platform)` tests, that must have the corresponding helpers available.

This will get cleaned-up later on, but rn a level of indirection is necessary (or alternatively add a bunch of stuff in the legacy files).

File: pkg/testutil/nerdtest/requirements.go (hunk 25 above)
Comment (apostasie) [reply pkg/testutil/nerdtest/requirements.go:135]: Yeah, I was wondering about that...

I am happy either way - I will take another look at that aspect (RootLess, RootFul) with the next round then.

File: pkg/testutil/nerdtest/utilities.go (hunk 26 above)
Comment (apostasie) [reply pkg/testutil/nerdtest/utilities.go:34]: Migrating to the proposed new framework, I would rather have everything sourced from there to help cleaning things up.
And yes, eventually, they will be merged.

**Comment by apostasie** [comment]
> Thanks, I'm merging this, but this amount of PR was almost unreviewable and even causing Safari to almost hang up 😓

@AkihiroSuda I really appreciate your help and support on this.
I could not have done it otherwise.

> In the next series of the PR please consider splitting the PR to smaller parts.

Upcoming parts are roughly:
- github actions clean-up with non-flaky/flaky(with retry) split up - this will give us a lot of bang for the buck and likely a good speed-up of the total testrun time
- migration of container tests
- migration of builder tests
- migration of login tests
- migration of compose tests

I think the test framework is mature enough now that we can have a separate PR for each of these ^ - making it much less painful for everybody (me included 😓).

> Also consider reducing the amount of the helper functions and constants (see the comments)

Eventually all of it will fuse into one - rn I have to make sure existing tests still work unaltered, and the new ones are not constrained... It is tough. I almost gave up on the effort multiple times 😰. Debugging some of these issues was mind altering.

Thanks again!
I am absolutely convinced we will be in a much better situation soon, and the quality of 2.0 will rock!
